test/cpp/api/support.h

```diff
@@ -61,5 +61,17 @@ inline int count_substr_occurrences(const std::string& str, const std::string& s
   return count;
 }
 
+// A RAII, thread local (!) guard that changes default dtype upon
```

Comment:
Just checking, `torch::set_default_dtype` changes the default dtype locally?

Reply:
It seems global: https://github.com/pytorch/pytorch/blob/master/c10/core/DefaultDtype.cpp. Orthogonal to this PR, does this need synchronization?

Reply:
Oh sorry, I totally did not realize this was in test code. This is fine in testing code, I was wondering if it was OK as an API.

Reply:
Thanks for the catch - I think this `AutoDefaultDtypeMode` guard in its original form was not safe to use if we ever run C++ tests in parallel in multiple threads, since a test in thread A might not have finished when another test in thread B changes the default dtype using `torch::set_default_dtype`. To address this problem, I made the following two changes:
1. Add `static std::mutex default_dtype_mutex` to `AutoDefaultDtypeMode`, so that usage of `AutoDefaultDtypeMode` is synchronized and thus always thread-safe.
2. Change other C++ tests that have `set_default_dtype(...)` to always use `AutoDefaultDtypeMode` to guard the whole scope of the test, so that access to `set_default_dtype(...)` is effectively synchronized *within the C++ test suite*. (It won't protect us against running a C++ test while changing default dtype through Python within the same process, but such usage seems more intentional than accidental, and the author of that test mechanism should be responsible for making sure such usage is sane.)

Reply:
I think the `AutoDefaultDtypeMode` guard is pretty ad-hoc - for now it's only used for the C++ tests, and its RAII mechanism should work :D

Reply:
Does gtest run tests in parallel / is it able to?

Reply:
I don't think gtest by itself can run tests in parallel (https://cuhkszlib-xiaoxing.readthedocs.io/en/latest/external/gtest/googletest/docs/FAQ.html#does-google-test-support-running-tests-in-parallel), but there are some test runners that can make it work: https://github.com/google/gtest-parallel

Reply:
Based on the design of the tests (they only run single-threaded), it might be better to simplify the data structure by removing the mutex and writing a quick comment that this isn't thread-safe and that it doesn't matter because we only use one thread